server/vehicle-persistence.lua

```diff
@@ -0,0 +1,108 @@
+if GetConvar('qbx:enable_vehicle_persistence', 'false') == 'false' then return end
+
+assert(lib.checkDependency('qbx_vehicles', '1.4.1', true))
+
+---A persisted vehicle will respawn when deleted. Only works for player owned vehicles.
+---Vehicles spawned using lib are automatically persisted
+---@param vehicle number
+local function enablePersistence(vehicle)
+    Entity(vehicle).state:set('persisted', true, true)
+end
+
+exports('EnablePersistence', enablePersistence)
+
+---A vehicle without persistence will not respawn when deleted.
+---@param vehicle number
+function DisablePersistence(vehicle)
+    Entity(vehicle).state:set('persisted', nil, true)
+end
+
+exports('DisablePersistence', DisablePersistence)
+
+local function getVehicleId(vehicle)
+    return Entity(vehicle).state.vehicleid or exports.qbx_vehicles:GetVehicleIdByPlate(GetVehicleNumberPlateText(vehicle))
+end
+
+RegisterNetEvent('qbx_core:server:vehiclePropsChanged', function(netId, diff)
+    local vehicle = NetworkGetEntityFromNetworkId(netId)
+
+    local vehicleId = getVehicleId(vehicle)
+    if not vehicleId then return end
+
+    local props = exports.qbx_vehicles:GetPlayerVehicle(vehicleId)?.props
+    if not props then return end
+
+    if diff.bodyHealth then
+        props.bodyHealth = GetVehicleBodyHealth(vehicle)
+    end
+
+    if diff.engineHealth then
+        props.engineHealth = GetVehicleEngineHealth(vehicle)
+    end
+
+    if diff.tankHealth then
+        props.tankHealth = GetVehiclePetrolTankHealth(vehicle)
+    end
+
+    if diff.fuelLevel then
+        props.fuelLevel = diff.fuelLevel ~= 'deleted' and diff.fuelLevel or nil
+    end
+
+    if diff.oilLevel then
+        props.oilLevel = diff.oilLevel ~= 'deleted' and diff.oilLevel or nil
+    end
+
+    if diff.dirtLevel then
+        props.dirtLevel = GetVehicleDirtLevel(vehicle)
+    end
+
+    if diff.windows then
+        props.windows = diff.windows ~= 'deleted' and diff.windows or nil
+    end
+
+    if diff.doors then
+        props.doors = diff.doors ~= 'deleted' and diff.doors or nil
+    end
+
+    if diff.tyres then
+        local damage = {}
+        for i = 0, 7 do
+            if IsVehicleTyreBurst(vehicle, i, false) then
+                damage[i] = IsVehicleTyreBurst(vehicle, i, true) and 2 or 1
+            end
+        end
+
+        props.tyres = damage
+    end
+
+    exports.qbx_vehicles:SaveVehicle(vehicle, {
+        props = props,
+    })
+end)
+
+AddEventHandler('qbx_core:server:vehicleSpawned', function(entity)
```

Comment:
What is this event for? Do we want it just to set `EnablePersistence` or is this for other resources to use too? If it's the former then we should just use `EnablePersistence`.

Reply:
I was thinking other resources, as one enabling persistence would want it enabled by default on all player vehicles.

Reply:
I'm failing to see a use case for this. Specifically for adding another handler to this event, since instead of triggering it you can just use the export.

Reply:
This is an internal event for core. Since lib.spawnVehicle is a library, when a vehicle is spawned an event is needed to tell core to persist the vehicle. Otherwise, callers would have to manually call the export after spawning a vehicle (is this what you are suggesting we do instead?)

Reply:
I'm saying wrapping it in an event feels unnecessary. Why not use the export directly in the lib?

Reply:
Switched to using the exports and got rid of the event. Make it so exports are always available, instead of only available if persistence is enabled.